core/modules/replica/StopRequest.cc

```diff
@@ -0,0 +1,205 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/StopRequest.h"
+
+// System headers
+#include <stdexcept>
+
+// Third party headers
+#include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "replica/ProtocolBuffer.h"
+#include "replica/ServiceProvider.h"
+
+// This macro to appear witin each block which requires thread safety
+#define LOCK_GUARD std::lock_guard<std::mutex> lock(_mtx)
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.StopRequest");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+StopRequestBaseM::StopRequestBaseM(ServiceProvider::pointer const& serviceProvider,
+                                   boost::asio::io_service& io_service,
+                                   char const*              requestTypeName,
+                                   std::string const&       worker,
+                                   std::string const&       targetRequestId,
+                                   proto::ReplicationReplicaRequestType requestType,
+                                   bool                     keepTracking,
+                                   std::shared_ptr<Messenger> const&  messenger)
+    :   RequestMessenger(serviceProvider,
+                         io_service,
+                         requestTypeName,
+                         worker,
+                         0,    /* priority */
+                         keepTracking,
+                         false /* allowDuplicate */,
+                         messenger),
+        _targetRequestId(targetRequestId),
+        _requestType(requestType) {
+}
+
+void StopRequestBaseM::startImpl() {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << "startImpl");
+
+    // Serialize the Request message header and the request itself into
+    // the network buffer.
+
```

Comment:
why no lock in `startImpl`? (Maybe a comment explaining why would be appropriate here?)

Reply:
This problem was gone after a major the thread-safety enforcement in this and other similar files.